### PR TITLE
Stop using dht::ring_position_view::max() in reader_selector

### PR DESCRIPTION
**Pull request by haaawk**
This is a first baby step in a long voyage to remove special MIN and MAX tokens from our codebase.

To be able to remove them we have to make sure that our code does not depend on them at all.
This means that transitively the code can't depend on:
ring_position::is_min()
ring_position::is_max()
ring_position_view::is_min()
ring_position_view::is_max()
ring_position_ext::is_min()
ring_position_ext::is_max()

This PR makes reader_selector and its subclasses independent from is_max() functions. dht::ring_position_view::max() was used to mark that the selector is finished but this can be done more explicitly by using optimized_optional<dht::ring_position_view> and setting it to disengaged optional when the selector is done.

Refs #1580 

Tests: unit(dev, release, debug), dtest(dev)

### CODE REVIEW

**Comment by haaawk** [comment]
BYO run is here: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1079/

**Comment by haaawk** [comment]
Rebased.

File: mutation_reader.hh [hunk 1]
```diff
@@ -35,20 +36,20 @@ class reader_selector {
 protected:
     schema_ptr _s;
 private:
-    dht::ring_position_view _selector_position;
+    optimized_optional<dht::ring_position_view> _selector_position;
 protected:
```
Comment (avikivity) [inline mutation_reader.hh:40]: I think this is introduces complications. ring_position_view should be able to represent the every bound that partition_range can represent.

One way to do that is to allow ring_position_view::_token to be null. Another way to to set it to the maximum numeric token (we can do that since we removed byte_ordered_partitioner), with _weight = 1. Probably the second way is better in terms of regression risk since it just sets the token to a value.

Comment (avikivity) [reply mutation_reader.hh:40]: Actually, a partition_range bound is similar to optional<ring_position>, so making ring_position optional is the right thing. It introduces complications but fewer in the long run. So I think this is the correct direction.

**Comment by avikivity** [comment]
Looks good. @denesb please review.

**Comment by tgrabiec** [comment]
What's the motivation for the change?

I think this is a step in the wrong direction. 

ring_position::max() and min() are needed to represent an arbitrary range, including an open-ended range, with a pair of position objects. Row cache is using it, for example, it can insert a dummy entry with the max() position to reflect an open-ended continuity.

An alternative is to represent it outside the object as optionals, but that's worse. We used to represent max/min bounds as optionals in the past in clustering_range, and it resulted in the need for special cases and interpretation in each user, which led to several bugs. Having max()/min()  part of the position object is better because the consumer doesn't have to handle special cases.

**Comment by haaawk** [comment]
> What's the motivation for the change?

Once we've fixed the token representation as int64_t we don't need special markers to represent MIN and MAX.
MIN is just std::numeric_limits<int64_t>::min() and MAX is just std::numeric_limits<int64_t>::max()

> 
> I think this is a step in the wrong direction.
> 
> ring_position::max() and min() are needed to represent an arbitrary range, including an open-ended range, with a pair of position objects. Row cache is using it, for example, it can insert a dummy entry with the max() position to reflect an open-ended continuity.

Wouldn't inserting dummy at std::numeric_limits<int64_t>::max() work as well?

> 
> An alternative is to represent it outside the object as optionals, but that's worse. We used to represent max/min bounds as optionals in the past in clustering_range, and it resulted in the need for special cases and interpretation in each user, which led to several bugs. Having max()/min() part of the position object is better because the consumer doesn't have to handle special cases.

I think we can still use std::numeric_limits<int64_t>::min() and std::numeric_limits<int64_t>::max() in cache, no?

**Comment by tgrabiec** [comment]
On Tue, Feb 23, 2021 at 11:57 AM Piotr Jastrzębski <notifications@github.com>
wrote:

> What's the motivation for the change?
>
> Once we've fixed the token representation as int64_t we don't need special
> markers to represent MIN and MAX.
> MIN is just std::numeric_limits<int64_t>::min() and MAX is just
> std::numeric_limits<int64_t>::max()
>
> I think this is a step in the wrong direction.
>
> ring_position::max() and min() are needed to represent an arbitrary range,
> including an open-ended range, with a pair of position objects. Row cache
> is using it, for example, it can insert a dummy entry with the max()
> position to reflect an open-ended continuity.
>
> Wouldn't inserting dummy at std::numeric_limits<int64_t>::max() work as
> well?
>
The position type is ring_position_view, not int64_t. So we need
ring_position::max(), which can use std::numeric_limits<int64_t>::max() for
the token part under the hood.

An alternative is to represent it outside the object as optionals, but

**Comment by tgrabiec** [comment]
That's assuming that no partition key can hash to std::numeric_limits<int64_t>::max(). Is it the case?

**Comment by haaawk** [comment]
> That's assuming that no partition key can hash to std::numeric_limits<int64_t>::max(). Is it the case?

Yes - I meant a ring_position(_view) that contains token with value std::numeric_limits<int64_t>::max().

Partition keys can be hashed to std::numeric_limits<int64_t>::max() but for dummy we can use ring_position_view with token of std::numeric_limits<int64_t>::max() value and _weight == 1.

BTW this is another redundancy of the existing code:
1. We can have MAX token that's represented by a special token_kind::after_all_keys
2. We can have ring_position(_view) with token equal to std::numeric_limits<int64_t>::max() and _weight == 1
They both mean very similar (if not exactly the same) thing.

**Comment by tgrabiec** [comment]
So to sum up: std::numeric_limits<int64_t>::max() can by used as a token value to implement ring_position_view::max(), but ring_position_view::max() should stay.

**Comment by haaawk** [comment]
> So to sum up: std::numeric_limits<int64_t>::max() can by used as a token value to implement ring_position_view::max(), but ring_position_view::max() should stay.

I guess so. My goal is not to remove it but to remove token_kind and make token=int64_t.

BTW it turns out that partition ranges that are just ranges of ring_position(_view)'s really don't need optional since we are able to represent -inf with ring_position(token = numeric_limits<int64_t>::min, weight=-1) and +inf with ring_position(token=numeric_limits<int64_t>::max(), weight=1).

So @avikivity my proposal would be to:
1. remove token_kind -> make token ~= int64_t
2. Change partition_range and any other range that's on ring_position(_view) to not have optionals as bounds

What do you think?

We probably could drop _inclusive from the interval_bound because inclusive/exclusive bound can be represented with ring_position o a proper weight. @tgrabiec what do you think?

To sum up,
token ~= int64_t
partition_range = pair<ring_position, ring_position>

**Comment by haaawk** [comment]
also maybe remove range::_singular

**Comment by denesb** [comment]
While we can drop the `_inclusive` flag from `nonwrapping_interval<dht::ring_position>`, we cannot currently drop it from `nonwrapping_interval<clustering_key>`, as the latter doesn't have said flag included. Maybe we should transition to `nonwrapping_interval<position_in_partition>` anyway.

**Comment by haaawk** [comment]
> While we can drop the `_inclusive` flag from `nonwrapping_interval<dht::ring_position>`, we cannot currently drop it from `nonwrapping_interval<clustering_key>`, as the latter doesn't have said flag included. Maybe we should transition to `nonwrapping_interval<position_in_partition>` anyway.

I agree. At the moment, I'm concentrating on partition range and tokens

**Comment by avikivity** [comment]
interval<> needs to have inclusive/exclusive support, since it can't count on the type T to be integral.

**Comment by avikivity** [comment]
> partition_range = pair<ring_position, ring_position>

This is a huge change. We lose all support for interval arithmetic.

**Comment by haaawk** [comment]
> > partition_range = pair<ring_position, ring_position>
> 
> This is a huge change. We lose all support for interval arithmetic.

We don't have to lose it. We can just adopt it.
The thing is that @tgrabiec is right that having a MIN and MAX ring_position is super convenient and we can already express all the inclusiveness/exclusiveness and -inf/+inf with just ring_position. Putting ring_position in an interval that duplicates all of this is error prone and confusing. And as @tgrabiec said, it leads to all sort of cumbersome checks when using pure ring_position would naturally handle all those cases with just a single comparator.

**Comment by avikivity** [comment]
I think ring_position has some overlap with interval<>, which would be good to remove.

But for removing token::min/max() that's not needed, since you can use min/max numeric tokens + token_bound::start/end.
